examples/simulation-tensorflow-tabular/sim.py

```diff
@@ -0,0 +1,249 @@
+        client_dataset = dataset.load_partition(int(cid))
+
+        # Now let's split it into train (90%) and validation (10%)
+        client_dataset_splits = client_dataset.train_test_split(test_size=0.1, seed=RANDOM_STATE)
+
+        trainset = client_dataset_splits["train"].to_tf_dataset(
+            columns="features", label_cols="labels", batch_size=64
+        )
+
+        valset = client_dataset_splits["test"].to_tf_dataset(
+            columns="features", label_cols="labels", batch_size=64
+        )
+
+        # Extract the number of features
+        element_spec = trainset.element_spec
+        num_features = element_spec[0].shape[1]
+
+        # Create and return client
+        return FlowerClient(trainset, valset, num_features).to_client()
+
+    return client_fn
+
+
+def weighted_average(metrics: List[Tuple[int, Metrics]]) -> Metrics:
+    """Aggregation function for (federated) evaluation metrics, i.e. those returned by
+    the client's evaluate() method."""
+    # Multiply accuracy of each client by number of examples used
+    accuracies = [num_examples * m["accuracy"] for num_examples, m in metrics]
+    examples = [num_examples for num_examples, _ in metrics]
+
+    # Aggregate and return custom metric (weighted average)
+    return {"accuracy": sum(accuracies) / sum(examples)}
+
+
+def get_evaluate_fn(testset: Dataset):
+    """Return an evaluation function for server-side (i.e. centralised) evaluation."""
+
+    # The `evaluate` function will be called after every round by the strategy
+    def evaluate(
+        server_round: int,
+        parameters: fl.common.NDArrays,
+        config: Dict[str, fl.common.Scalar],
+    ):
+        # Extract the number of features
+        element_spec = testset.element_spec
+        num_features = element_spec[0].shape[1]
+
+        model = get_model(num_features)  # Construct the model
+        model.set_weights(parameters)  # Update model with the latest parameters
+        loss, accuracy = model.evaluate(testset, verbose=VERBOSE)
+        return loss, {"accuracy": accuracy}
+
+    return evaluate
+
+# Preprocess the data
+def preprocess_data(data):
+    """Preprocess the dataset."""
+    # Select relevant features
+    features = data[RELEVANT_FEATURES].copy()
+    
+    # Convert 'Sex' to binary values
+    features['Sex'] = features['Sex'].map({'male': 0, 'female': 1})
+    
+    # Fill missing 'Age' values with the mean age
+    val_miss_age = features['Age'].mean()
+    features['Age'].fillna(val_miss_age, inplace=True)
+    
+    # Scale the features
+    scaler = StandardScaler()
+    features = pd.DataFrame(scaler.fit_transform(features), columns=RELEVANT_FEATURES)
+
+    data = pd.concat([features,data['Survived']], axis=1)
+    
+    return data
+
+# Load dataset from Hugging Face
+titanic_cds = load_dataset("julien-c/titanic-survival")
+
+# Convert to pandas to preprocess
+titanic_cds = titanic_cds['train'].to_pandas()
+titanic_cds = preprocess_data(titanic_cds)
```

Comment:
That's a leak in the preprocessing dataset. Each partition should be normalized/standardized according to ONLY the LOCAL data.

Reply:
Thanks! I modified the preprocessing step to be done ONLY with the LOCAL data.

PS: Check the new [PR #3633](https://github.com/adap/flower/pull/3633) with all the modifications.